packages/wonder-blocks-dropdown/components/action-menu.js

```diff
@@ -0,0 +1,237 @@
+
+import type {
+    ActionItemProps,
+    SelectItemProps,
+    SeparatorProps,
+} from "../utils/types.js";
+
+type OpenerProps = {
+    /**
+     * Display text for the opener.
+     */
+    children: string,
+    /**
+     * Whether the opener is disabled. If disabled, disallows interaction.
+     */
+    disabled?: boolean,
+    /**
+     * Callback for when the opener is pressed.
+     */
+    onClick: () => void,
+};
+
+class ActionMenuOpener extends React.Component<OpenerProps> {
+    static defaultProps = {
+        disabled: false,
+    };
+
+    render() {
+        const {children, disabled, onClick} = this.props;
+
+        // TODO: incorporate caret once Icon is done
+        return (
+            <Button
+                role="menu"
+                color={"default"}
+                disabled={disabled}
+                kind={"tertiary"}
+                light={false}
+                onClick={onClick}
+                style={[styles.opener]}
+            >
+                {children}
+            </Button>
+        );
+    }
+}
+
+type ItemProps = ActionItemProps | SelectItemProps | SeparatorProps;
+
+type MenuProps = {
+    /**
+     * The items in this menu.
+     */
+    items: Array<ItemProps>,
+
+    /**
+     * Text for the opener of this menu.
+     */
+    menuText: string,
+
+    /**
+     * Whether this menu contains select items. If true, user should also define
+     * the onChange and selectedItems props.
+     */
+    containsSelectionItems?: boolean,
+
+    /**
+     * A callback that returns items that are newly selected. Use only if this
+     * menu contains select items (prop containsSelectionItems).
+     */
+    onChange?: (selectedItems: Array<string>) => void,
+
+    /**
+     * The values of the items that are currently selected. Use only if this
+     * menu contains select items (prop containsSelectionItems).
+     */
+    selectedValues?: Array<string>,
+
+    /**
+     * Whether this menu should be left-aligned or right-aligned with the
+     * opener component. Defaults to left-aligned.
+     */
+    alignment?: "left" | "right",
+
+    /**
+     * Whether this menu is disabled. A disabled menu may not be opened and
+     * does not support interaction. Defaults to false.
+     */
+    disabled?: boolean,
+
+    /**
+     * Optional styling to add.
+     */
+    style?: any,
+};
+
+type State = {
+    /**
+     * Whether or not menu is open.
+     */
+    open: boolean,
+};
+
+export default class ActionMenu extends React.Component<MenuProps, State> {
+    static defaultProps = {
+        alignment: "left",
+        containsSelectionItems: false,
+        disabled: false,
+    };
+
+    constructor(props: MenuProps) {
+        super(props);
+
+        this.state = {
+            open: false,
+        };
+    }
+
+    toggleMenu() {
+        this.setState((prevState) => ({
+            open: !prevState.open,
+        }));
+    }
+
+    handleSelected(selectedValue: string, newSelectionState: boolean) {
```

Comment:
Could this be a toggle and then we wouldn't need the `newSelectionState` or are there cases where that isn't sufficient?

Reply:
I changed `SelectItem` to pass back the old selection state instead. Technically, the menu has this information (but it would have to filter for the value). I think this information is useful because in the case of the single select menu, clicking on a selected item doesn't toggle it. So this signature from `SelectItem` allows each menu to decide what to do.